File: ga/4omini_result_2024_7/result/queue_mini/38_queue.c
use std::io::{self, Write};
use std::ops::{Deref, DerefMut};

type ElementType = i32;

#[derive(Debug)]
struct Queue {
    capacity: usize,
    front: usize,
    rear: isize,
    size: usize,
    array: Vec<ElementType>,
}

impl Queue {
    fn new(max_elements: usize) -> Result<Self, String> {
        if max_elements < 5 {
            return Err("Queue size is too small".to_string());
        }
        let array = vec![0; max_elements];
        Ok(Queue {
            capacity: max_elements,
            front: 0,
            rear: -1,
            size: 0,
            array,
        })
    }

    fn is_empty(&self) -> bool {
        self.size == 0
    }

    fn is_full(&self) -> bool {
        self.size == self.capacity
    }

    fn make_empty(&mut self) {
        self.size = 0;
        self.front = 0;
        self.rear = -1;
    }

    fn enqueue(&mut self, x: ElementType) -> Result<(), String> {
        if self.is_full() {
            return Err("Full queue".to_string());
        }
        self.size += 1;
        self.rear = self.successor(self.rear);
        self.array[self.rear as usize] = x;
        Ok(())
    }

    fn front(&self) -> Result<ElementType, String> {
        if self.is_empty() {
            return Err("Empty queue".to_string());
        }
        Ok(self.array[self.front])
    }

    fn dequeue(&mut self) -> Result<(), String> {
        if self.is_empty() {
            return Err("Empty queue".to_string());
        }
        self.size -= 1;
        self.front = self.successor(self.front as isize) as usize;
        Ok(())
    }

    fn front_and_dequeue(&mut self) -> Result<ElementType, String> {
        let x = self.front()?;
        self.size -= 1;
        self.front = self.successor(self.front as isize) as usize;
        Ok(x)
    }

    fn successor(&self, value: isize) -> isize {
        if value + 1 == self.capacity as isize {
            0
        } else {
            value + 1
        }
    }
}

fn main() {
    let max_elements = 12;
    let mut q = Queue::new(max_elements).expect("Failed to create queue.");

    for i in 0..10 {
        q.enqueue(i).expect("Failed to enqueue.");
    }

    while !q.is_empty() {
        println!("{}", q.front().expect("Failed to read front."));
        q.dequeue().expect("Failed to dequeue.");
    }

    for i in 0..10 {
        q.enqueue(i).expect("Failed to enqueue.");
    }

    while !q.is_empty() {
        println!("{}", q.front().expect("Failed to read front."));
        q.dequeue().expect("Failed to dequeue.");
    }
}

File: ga/gpt3_5_result_2024_6/result/queue_gpt/38_queue.rs
use std::ptr;

struct QueueRecord {
    capacity: usize,
    front: usize,
    rear: usize,
    size: usize,
    array: Vec<ElementType>,
}

type Queue = Box<QueueRecord>;

type ElementType = i32;

const MIN_QUEUE_SIZE: usize = 5;

fn is_empty(Q: &Queue) -> bool {
    Q.size == 0
}

fn is_full(Q: &Queue) -> bool {
    Q.size == Q.capacity
}

fn create_queue(max_elements: usize) -> Queue {
    if max_elements < MIN_QUEUE_SIZE {
        panic!("Queue size is too small");
    }

    let mut array = Vec::with_capacity(max_elements);
    unsafe { array.set_len(max_elements) };

    Box::new(QueueRecord {
        capacity: max_elements,
        front: 1,
        rear: 0,
        size: 0,
        array,
    })
}

fn make_empty(Q: &mut Queue) {
    Q.size = 0;
    Q.front = 1;
    Q.rear = 0;
}

fn dispose_queue(Q: Queue) {
    drop(Q);
}

fn succ(value: usize, Q: &Queue) -> usize {
    if value + 1 == Q.capacity {
        0
    } else {
        value + 1
    }
}

fn enqueue(X: ElementType, Q: &mut Queue) {
    if is_full(Q) {
        panic!("Full queue");
    } else {
        Q.size += 1;
        Q.rear = succ(Q.rear, Q);
        Q.array[Q.rear] = X;
    }
}

fn front(Q: &Queue) -> ElementType {
    if !is_empty(Q) {
        Q.array[Q.front]
    } else {
        panic!("Empty queue");
    }
}

fn dequeue(Q: &mut Queue) {
    if is_empty(Q) {
        panic!("Empty queue");
    } else {
        Q.size -= 1;
        Q.front = succ(Q.front, Q);
    }
}

fn front_and_dequeue(Q: &mut Queue) -> ElementType {
    if is_empty(Q) {
        panic!("Empty queue");
    } else {
        let X = Q.array[Q.front];
        Q.size -= 1;
        Q.front = succ(Q.front, Q);
        X
    }
}

fn main() {
    let mut Q = create_queue(12);

    for i in 0..10 {
        enqueue(i, &mut Q);
    }

    while !is_empty(&Q) {
        println!("{}", front(&Q));
        dequeue(&mut Q);
    }

    for i in 0..10 {
        enqueue(i, &mut Q);
    }

    while !is_empty(&Q) {
        println!("{}", front(&Q));
        dequeue(&mut Q);
    }

    dispose_queue(Q);
}